cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/NotificationSubscriberService.java

```diff
@@ -0,0 +1,341 @@
+                                ProgramLifecycleService lifecycleService, PropertiesResolver propertiesResolver,
+                                NamespaceQueryAdmin namespaceQueryAdmin,
+                                CConfiguration cConf,
+                                DatasetFramework datasetFramework,
+                                TransactionSystemClient txClient) {
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.store = store;
+    this.lifecycleService = lifecycleService;
+    this.propertiesResolver = propertiesResolver;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.messagingService = messagingService;
+    this.topics = cConf.getTrimmedStrings(Constants.AppFabric.SCHEDULER_MESSAGING_TOPICS);
+    this.subscriberThreads = new ArrayList<>();
+    this.readyJobs = new LinkedBlockingQueue<>();
+    this.runFlag = new AtomicBoolean();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("notification-subscriber-task")));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+  }
+
+  /**
+   * Temporary workaround when Schedule store is not ready.
+   *
+   * @param scheduleMap Map storing keys constructed from {@link Notification}
+   *                    and corresponding {@link ProgramSchedule} as values
+   */
+  public void setScheduleMap(Map<String, List<ProgramSchedule>> scheduleMap) {
+    this.scheduleMap = scheduleMap;
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running NotificationSubscriberService");
+    if (!isRunning()) {
+      return;
+    }
+
+    for (String topic : topics) {
+      subscriberThreads.add(new NotificationSubscriberThread(topic, null));
+    }
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.start();
+    }
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      try {
+        thread.join();
+      } catch (InterruptedException e) {
+        LOG.info("Thread {} is being terminated while waiting for it to finish.", thread.getName());
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    LOG.info("Stopping NotificationSubscriberService.");
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.interrupt();
+    }
+    if (taskExecutorService != null) {
+      taskExecutorService.shutdownNow();
```

Comment:
Usually the shutdown of executor is done in `shutdown` instead of `triggerShutdown`.